### PR TITLE
Add custom 404 page in documentation

### LINKED ISSUE #1964: Add a custom 404 page to docs site
## Problem

The doc's site 404 is the generic GitHub Pages 404, which is not helpful.

## Description

Create a simply 404.html for the docs site that renders the toctree with a note that the page requested could not be found.

## Additional context

Marked as high due to the docs reorg and the number of 404s we'll get as a result.

### PR DESCRIPTION
**Pull request by ritesh-pandey**
The custom 404 page will override the default 404 page rendered by GH Pages

## Fixes

Fixes #1964 by @sarayourfriend 

## Description

Add a custom 404.html page to override default 404 GH Page.

Screenshot of page.
image

## Checklist

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: documentation/conf.py [hunk 1]
```diff
@@ -34,6 +34,7 @@ def add_ext_to_path():
     "sphinx_copybutton",
 ]
 myst_heading_anchors = 6  # Add anchors to all headers, this is disabled by default.
+myst_enable_extensions = ["fieldlist"]
```
Comment (AetherUnbound) [inline documentation/conf.py:37]: Do you mind adding a comment on why this was added?

**Review by AetherUnbound — CHANGES_REQUESTED** [review]
Thanks for the contribution @ritesh-pandey! When I try to visit a missing page locally though, I still see the base 404 page (example with http://0.0.0.0:50231/ingestion_server/guides/foobar.html):

image

It seems like we may need to add the [`sphinx-notfound-page` plugin](https://sphinx-notfound-page.readthedocs.io/en/latest/configuration.html) to accomplish this, is that something you'd feel comfortable attempting?

File: documentation/conf.py (hunk 1 above)
Comment (ritesh-pandey) [reply documentation/conf.py:37]: We need to use `:orphan:` field in 404 page to suppress TOCTREE warning. Sphinx warns if a page is not included in TOC.   

[Field lists](https://www.sphinx-doc.org/en/master/usage/restructuredtext/field-lists.html) are not enabled by default in MyST. As per [MyST doc](https://myst-parser.readthedocs.io/en/latest/syntax/optional.html#field-lists), we need to add the mentioned line in conf file to enable fields.

**Comment by ritesh-pandey** [comment]
[GH Pages has capability to render a custom 404 page](https://docs.github.com/en/pages/getting-started-with-github-pages/creating-a-custom-404-page-for-your-github-pages-site) if it finds a `404.html`/`404.md` file present at root directory. GH Pages will automatically render the custom 404 page if we try to access a non-existent page.

Similar functionality is not available in Shpinx development server that we run locally. So, to test appearance of 404 page in local server, we need to access it directly i.e. http://0.0.0.0:50231/404.html 

I experimented with `sphinx-notfound-page` plugin and I can implement it. But it will not solve the problem of **404 page not getting rendered on local server**. [FAQ](https://sphinx-notfound-page.readthedocs.io/en/latest/faq.html#why-is-my-local-web-server-not-showing-a-404-html) in the extension talks about it.

File: documentation/404.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+:orphan:
```
Comment (dhruvkb) [inline documentation/404.md:1]: Going by how the other directives work, I thought maybe

```{orphan}
```

was the Markdown corresponding to the `:orphan:` in rST.

**Review by dhruvkb — APPROVED** [review]
GitHub Pages specifically uses a `404.html` page if no page matches the path. So this should work 👍, LGTM!

File: documentation/conf.py (hunk 1 above)
Comment (AetherUnbound) [inline documentation/conf.py:37]:
```suggestion
myst_enable_extensions = ["fieldlist"]  # Required for specifying :orphan: pages
```

Comment (AetherUnbound) [reply documentation/conf.py:37]: Got it, this is helpful thank you!

**Review by AetherUnbound — APPROVED** [review]
Thanks so much for the additional context on this @ritesh-pandey! I was unaware of the auto-magic work GitHub was doing to get 404s working, I appreciate you linking that 😄 One small note to add a comment (which I can commit), then this can be merged!

**Comment by dhruvkb** [comment]
Updated branch with `main` to see if CI passes.

**Pull request merged**